lib/prelude.dx

```diff
@@ -983,6 +990,32 @@ instance Monoid (List a)
         True  -> xs.(unsafeFromOrdinal _ i')
         False -> ys.(unsafeFromOrdinal _ (i' - nx))
 
+def ListMonoid (a:Type) : Monoid (List a) =
+  A = a -- XXX: Typing `Monoid a` below would quantify it over a,
+        --      which we don't want.
+  named-instance result : Monoid (List A)
+    mempty = mempty
+    mcombine = mcombine
+  result
```

Comment:
This is an interesting construction. Is the idea that you want to explicitly pass `a:Type` as an argument, and regular named instances are always implicit? (And is that because type inference isn't powerful enough to figure it out on its own?)

Reply:
I appreciate your tact in calling this `interesting`.  I have to admit to cargo culting here (with @dougalm 's help), copying similar constructions from the other typelcass-indexed monoids.  But yes, as far as I understand, this is a kludge that's necessary until type inference is more powerful.  At least I tried using `_` in my examples and the compiler said it couldn't synthesize an instance.

Reply:
Yeah are there other examples of this? I couldn't figure out what was going on, but it seems cool. What specifically is it fixing?

Reply:
Yes, for the [AddMonoid](https://github.com/google-research/dex-lang/blob/main/lib/prelude.dx#L293) and [MulMonoid](https://github.com/google-research/dex-lang/blob/main/lib/prelude.dx#L300).  The one I wrote is especially gory since it's just wrapping the existing [List Monoid](https://github.com/google-research/dex-lang/blob/main/lib/prelude.dx#L974). 
 But my understanding is that this can all be cleaned up once the type inference is better.  In the meantime I was parallelizing things in my demos and thought I'd share what I have so far.
 
 BTW @srush I just saw your discussion where your `filter` function takes a list instead of a table.  I'm not sure which way is better.

Reply:
Huh, I think I kind of get it. I need to learn a bit more about the type inference system to get why this fails and the table one succeeds. I guess it has to do with `a` being unspecified?

Btw, re:List / Table. I wrote up my thoughts here: (https://github.com/google-research/dex-lang/discussions/484). I find this distinction comes up a lot. As a user, my preference would be for functions to default to List if they are agnostic to the index type. Currently that is a bit annoying, but that seems more correct to me? 

In practice this seems kind of messy. Like the fact that [1, 2] <> [3, 4] => [4, 6] (I think?) but (toList [1, 2]) <> (toList [3, 4]) => AsList 4 [1, 2, 3, 4].

Reply:
Ugh, this is probably necessary because type inference really dislikes the type of the `Accum` handlers. In particular, the only thing that relates the monoid type and the accumulator type is the `AccumMonoid` class requirement, but those requirements are not used during type inference at the moment. So right now there is _nothing_ that allows the inference of monoids.

One way to make that better would be to have weaker (stronger for type inference) overloads of `Accum` functions such as:
```
yieldAccum (m : Monoid w) (action : h:Type ?-> AccumMonoid h w -> Ref h w -> {Writer} a) : a
``` 
This would clearly indicate the the desired monoid type has to be _exactly_ whatever type was inferred for the reference.

Reply:
@srush the reason why we don't want to default to `List` is that `List`s are notoriously difficult to generate code for, at least compared to indexed tables. Doing optimizations is much easier when you know more about the shape of your data.

Reply:
And yeah, we should document the type-class inference better. The gist of it is that it tries to synthesize the right type based on the bindings in scope that are additionally marked as _instance bindings_. Then, `instance Class Type` is syntactic sugar for an instance binding that creates the class dictionary. `named-instance` also constructs the class dictionary and allows you to give it a name, but it doesn't mark the binding (to the given name) as an instance binding, so by default it does not participate in the search (unless marked with `%instance` at some point).